### ISSUE TITLE
NCCL 2.17.1-1 Build error on debian 10

### ISSUE THREAD

**Issue by hanzhi713** [NVIDIA/nccl#835]
Since NCCL 2.17.1 is not available pre-built for debian 10 and ubuntu 18, I'm trying to build it from source.

cuda version: 12.1
gcc version: 8.3.0
build command: 
```
git clone -b v2.17.1-1 https://github.com/NVIDIA/nccl
cd nccl
make -j1 src.build NVCC_GENCODE="-gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_90,code=sm_90"
```

Error message:

```
make -C src build BUILDDIR=/nccl/build
make[1]: Entering directory '/nccl/src'
NVCC_GENCODE is -gencode=arch=compute_70,code=sm_70 -gencode=arch=compute_80,code=sm_80 -gencode=arch=compute_90,code=sm_90 -gencode=arch=compute_90,code=compute_90
Generating nccl.h.in                           > /nccl/build/include/nccl.h
Grabbing   include/nccl_net.h                  > /nccl/build/include/nccl_net.h
Compiling  init.cc                             > /nccl/build/obj/init.o
In file included from include/nvtx.h:10,
                 from include/core.h:62,
                 from include/info.h:13,
                 from include/graph.h:112,
                 from include/transport.h:11,
                 from include/comm.h:10,
                 from include/channel.h:9,
                 from init.cc:8:
include/nvtx3/nvtx3.hpp: In instantiation of ‘struct nvtx3::v1::detail::has_name<nvtx3::v1::domain::global, void>’:
include/nvtx3/nvtx3.hpp:830:7:   required by substitution of ‘template<class D, typename std::enable_if<(! nvtx3::v1::detail::has_name<D>::value), int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get() [with D = nvtx3::v1::domain::global; typename std::enable_if<(! nvtx3::v1::detail::has_name<D>::value), int>::type <anonymous> = <missing>]’
include/nvtx3/nvtx3.hpp:942:52:   required from here
include/nvtx3/nvtx3.hpp:651:29: error: ‘name’ is not a member of ‘nvtx3::v1::domain::global’
 struct has_name<T, decltype((void)T::name, void())> : std::true_type {};
                             ^~~~~~~
include/nvtx3/nvtx3.hpp:651:29: error: ‘name’ is not a member of ‘nvtx3::v1::domain::global’
include/nvtx3/nvtx3.hpp:942:22: error: template-id ‘get<nvtx3::v1::domain::global>’ for ‘const nvtx3::v1::domain& nvtx3::v1::domain::get()’ does not match any template declaration
 inline domain const& domain::get<domain::global>() noexcept
                      ^~~~~~
include/nvtx3/nvtx3.hpp:832:24: note: candidates are: ‘template<class D, typename std::enable_if<(! nvtx3::v1::detail::has_name<D>::value), int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:814:24: note:                 ‘template<class D, typename std::enable_if<(! nvtx3::v1::detail::is_c_string<decltype (D::name)>::value), int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:799:24: note:                 ‘template<class D, typename std::enable_if<nvtx3::v1::detail::is_c_string<decltype (D::name)>::value, int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp: In instantiation of ‘nvtx3::v1::range_handle nvtx3::v1::start_range_in(const nvtx3::v1::event_attributes&) [with D = nvtx3::v1::domain::global]’:
include/nvtx3/nvtx3.hpp:2399:45:   required from here
include/nvtx3/nvtx3.hpp:2323:60: error: no matching function for call to ‘nvtx3::v1::domain::get<nvtx3::v1::domain::global>()’
   return range_handle{nvtxDomainRangeStartEx(domain::get<D>(), attr.get())};
                                              ~~~~~~~~~~~~~~^~
include/nvtx3/nvtx3.hpp:799:24: note: candidate: ‘template<class D, typename std::enable_if<nvtx3::v1::detail::is_c_string<decltype (D::name)>::value, int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:799:24: note:   template argument deduction/substitution failed:
include/nvtx3/nvtx3.hpp:798:20: error: ‘name’ is not a member of ‘nvtx3::v1::domain::global’
     , int>::type = 0>
                    ^
include/nvtx3/nvtx3.hpp:814:24: note: candidate: ‘template<class D, typename std::enable_if<(! nvtx3::v1::detail::is_c_string<decltype (D::name)>::value), int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:814:24: note:   template argument deduction/substitution failed:
include/nvtx3/nvtx3.hpp:812:7: error: ‘name’ is not a member of ‘nvtx3::v1::domain::global’
       !detail::is_c_string<decltype(D::name)>::value
       ^~~~~~~
include/nvtx3/nvtx3.hpp:832:24: note: candidate: ‘template<class D, typename std::enable_if<(! nvtx3::v1::detail::has_name<D>::value), int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:832:24: note:   template argument deduction/substitution failed:
include/nvtx3/nvtx3.hpp:831:20: error: no type named ‘type’ in ‘struct std::enable_if<false, int>’
     , int>::type = 0>
                    ^
include/nvtx3/nvtx3.hpp:2323:75: error: no matching function for call to ‘nvtx3::v1::range_handle::range_handle(<brace-enclosed initializer list>)’
   return range_handle{nvtxDomainRangeStartEx(domain::get<D>(), attr.get())};
                                                                           ^
include/nvtx3/nvtx3.hpp:2255:13: note: candidate: ‘constexpr nvtx3::v1::range_handle::range_handle(std::nullptr_t)’
   constexpr range_handle(std::nullptr_t) noexcept {}
             ^~~~~~~~~~~~
include/nvtx3/nvtx3.hpp:2255:13: note:   conversion of argument 1 would be ill-formed:
include/nvtx3/nvtx3.hpp:2234:13: note: candidate: ‘constexpr nvtx3::v1::range_handle::range_handle()’
   constexpr range_handle() noexcept = default;
             ^~~~~~~~~~~~
include/nvtx3/nvtx3.hpp:2234:13: note:   candidate expects 0 arguments, 1 provided
include/nvtx3/nvtx3.hpp:2225:22: note: candidate: ‘constexpr nvtx3::v1::range_handle::range_handle(nvtx3::v1::range_handle::value_type)’
   constexpr explicit range_handle(value_type id) noexcept : _range_id{id} {}
                      ^~~~~~~~~~~~
include/nvtx3/nvtx3.hpp:2225:22: note:   conversion of argument 1 would be ill-formed:
include/nvtx3/nvtx3.hpp:2216:8: note: candidate: ‘constexpr nvtx3::v1::range_handle::range_handle(const nvtx3::v1::range_handle&)’
 struct range_handle {
        ^~~~~~~~~~~~
include/nvtx3/nvtx3.hpp:2216:8: note:   conversion of argument 1 would be ill-formed:
include/nvtx3/nvtx3.hpp:2216:8: note: candidate: ‘constexpr nvtx3::v1::range_handle::range_handle(nvtx3::v1::range_handle&&)’
include/nvtx3/nvtx3.hpp:2216:8: note:   conversion of argument 1 would be ill-formed:
include/nvtx3/nvtx3.hpp: In instantiation of ‘void nvtx3::v1::end_range_in(nvtx3::v1::range_handle) [with D = nvtx3::v1::domain::global]’:
include/nvtx3/nvtx3.hpp:2484:33:   required from here
include/nvtx3/nvtx3.hpp:2462:36: error: no matching function for call to ‘nvtx3::v1::domain::get<nvtx3::v1::domain::global>()’
   nvtxDomainRangeEnd(domain::get<D>(), r.get_value());
                      ~~~~~~~~~~~~~~^~
include/nvtx3/nvtx3.hpp:799:24: note: candidate: ‘template<class D, typename std::enable_if<nvtx3::v1::detail::is_c_string<decltype (D::name)>::value, int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:799:24: note:   template argument deduction/substitution failed:
include/nvtx3/nvtx3.hpp:798:20: error: ‘name’ is not a member of ‘nvtx3::v1::domain::global’
     , int>::type = 0>
                    ^
include/nvtx3/nvtx3.hpp:814:24: note: candidate: ‘template<class D, typename std::enable_if<(! nvtx3::v1::detail::is_c_string<decltype (D::name)>::value), int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:814:24: note:   template argument deduction/substitution failed:
include/nvtx3/nvtx3.hpp:812:7: error: ‘name’ is not a member of ‘nvtx3::v1::domain::global’
       !detail::is_c_string<decltype(D::name)>::value
       ^~~~~~~
include/nvtx3/nvtx3.hpp:832:24: note: candidate: ‘template<class D, typename std::enable_if<(! nvtx3::v1::detail::has_name<D>::value), int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:832:24: note:   template argument deduction/substitution failed:
include/nvtx3/nvtx3.hpp:831:20: error: no type named ‘type’ in ‘struct std::enable_if<false, int>’
     , int>::type = 0>
                    ^
include/nvtx3/nvtx3.hpp: In instantiation of ‘void nvtx3::v1::mark_in(const nvtx3::v1::event_attributes&) [with D = nvtx3::v1::domain::global]’:
include/nvtx3/nvtx3.hpp:2702:31:   required from here
include/nvtx3/nvtx3.hpp:2638:34: error: no matching function for call to ‘nvtx3::v1::domain::get<nvtx3::v1::domain::global>()’
   nvtxDomainMarkEx(domain::get<D>(), attr.get());
                    ~~~~~~~~~~~~~~^~
include/nvtx3/nvtx3.hpp:799:24: note: candidate: ‘template<class D, typename std::enable_if<nvtx3::v1::detail::is_c_string<decltype (D::name)>::value, int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:799:24: note:   template argument deduction/substitution failed:
include/nvtx3/nvtx3.hpp:798:20: error: ‘name’ is not a member of ‘nvtx3::v1::domain::global’
     , int>::type = 0>
                    ^
include/nvtx3/nvtx3.hpp:814:24: note: candidate: ‘template<class D, typename std::enable_if<(! nvtx3::v1::detail::is_c_string<decltype (D::name)>::value), int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:814:24: note:   template argument deduction/substitution failed:
include/nvtx3/nvtx3.hpp:812:7: error: ‘name’ is not a member of ‘nvtx3::v1::domain::global’
       !detail::is_c_string<decltype(D::name)>::value
       ^~~~~~~
include/nvtx3/nvtx3.hpp:832:24: note: candidate: ‘template<class D, typename std::enable_if<(! nvtx3::v1::detail::has_name<D>::value), int>::type <anonymous> > static const nvtx3::v1::domain& nvtx3::v1::domain::get()’
   static domain const& get() noexcept
                        ^~~
include/nvtx3/nvtx3.hpp:832:24: note:   template argument deduction/substitution failed:
include/nvtx3/nvtx3.hpp:831:20: error: no type named ‘type’ in ‘struct std::enable_if<false, int>’
     , int>::type = 0>
                    ^
make[1]: *** [Makefile:121: /nccl/build/obj/init.o] Error 1
make[1]: Leaving directory '/nccl/src'
make: *** [Makefile:25: src.build] Error 2
```

Upon closer inspection, the error may come from the fact that `struct global` has no member called `name`. 

https://github.com/NVIDIA/nccl/blob/9b7d5edbfcd70a1ab0874e27382d5c31f769aa0f/src/include/nvtx3/nvtx3.hpp#L747-L748

I'm able to solve the problem by setting `NVTX3_USE_CHECKED_OVERLOADS_FOR_GET=0` in CXXFLAGS, but I'm not quite sure why this error even exists since NCCL 2.17.1 was successfully built and released by the Nvidia team.

**Comment by svavil** [comment]
@hanzhi713 I'm trying to build from source another program that uses `NVTX`, and I'm running into similar issues during compilation. Can you describe in more details how you set the `NVTX3_USE_CHECKED_OVERLOADS_FOR_GET` flag? Thank you.

**Comment by hanzhi713** [comment]
> NVTX3_USE_CHECKED_OVERLOADS_FOR_GET

Basically you want to set the macro `NVTX3_USE_CHECKED_OVERLOADS_FOR_GET` to `0` during compilation. For NCCL build, you can add `-DNVTX3_USE_CHECKED_OVERLOADS_FOR_GET=0` to `CXXFLAGS` to `makefiles/common.mk`

```
CXXFLAGS += -DNVTX3_USE_CHECKED_OVERLOADS_FOR_GET=0
```

**Comment by AddyLaddy** [comment]
Does this failure still occur with the NVTX headers in NCCL 2.18.1 ?

**Comment by toylsc** [comment]
same problem when building with gcc8 in centos7, while gcc7 is ok.

**Comment by toylsc** [comment]
> @AddyLaddy same problem when building with gcc8 in centos7, while gcc7 is ok.

tested nccl version is 2.18.1-1

**Comment by Jackwin** [comment]
Hi all, is this issue fixed? the same issue occurs in nccl 2.18 using debian 10, cuda 12.1.

**Comment by jcohen-nvidia** [comment]
I'm looking into it.  The issue seems to be that the compiler version being used is having trouble with SFINAE.  I specifically made the NVTX3_USE_CHECKED_OVERLOADS_FOR_GET option for bypassing the extra sanity-checking code that uses SFINAE tricks because VS2015's compiler couldn't handle it.  But I tested every version of GCC from 4.8 to present and clang from 3.8 to present, and never had this problem with any of them.  Now, it seems you're not the only people hitting this -- some folks on QNX are seeing it as well.  I can't repro the issue locally with NCCL so I've been trying to get the updated license set up for QNX cross-compiling on the newest toolchain, and that took a full week, unfortunately.  Hopefully I can figure out a clean workaround for this.  It may just be adding a few more compilers to the list that disable NVTX3_USE_CHECKED_OVERLOADS_FOR_GET, but I want to understand why anyone hits this with modern GCC versions while my all-compiler tests never do.

**Comment by jcohen-nvidia** [comment]
Just a quick clarification:  Is it true that you only hit this with GCC 8.x, but do not hit it with older or newer GCC versions?  And if that's true can you check when minor versions of GCC 8 you're seeing the problem on?

**Comment by toylsc** [comment]
> Just a quick clarification: Is it true that you only hit this with GCC 8.x, but do not hit it with older or newer GCC versions? And if that's true can you check when minor versions of GCC 8 you're seeing the problem on?
> 
> I found this GCC bug report that shows some previously-working SFINAE cases were broken in GCC 8 up until this was fixed in 8.4: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=87748 It looks like I was testing on 8 with 8.4 most recently, so I will try to get older 8.x versions to test. Also, if it's easy for you to try updating GCC to 8.4 or 8.5 and seeing if the error goes away, that would be very helpful for me!

After testing different GCC versions in CentOS 7 (through devtoolset), I have confirmed that GCC 7.3.1 and 9.3.1 could successfully compile NCCL 2.18.1 while GCC 8.3.1 couldn't. I think it's true that the cause is the bug in GCC 8.3.1. Since devtoolset doesn't provide GCC 8.4, I have no environment for testing GCC 8.4

**Comment by jcohen-nvidia** [comment]
I'm in the process of doing more rigorous testing, but I have a workaround that you can try that should work.  It is basically the same workaround you did above to set NVTX3_USE_CHECKED_OVERLOADS_FOR_GET to 0, but this is a change that does it automatically for the compilers that are known to have broken SFINAE support.  So on line 565 or so of nvtx3.hpp, I will replace this:

```
#if defined(_MSC_VER) && _MSC_VER < 1914
```
with this:
```
#if defined(_MSC_VER) && _MSC_VER < 1914 \
  || defined(__GNUC__) && __GNUC__ == 8 && __GNUC_MINOR__ < 4
```
That should cover this range of 8.1-8.3 where things were broken, so you don't have to do the workaround `#define` yourself.  Give that nvtx3.hpp change a try locally if you'd like.

**Comment by AddyLaddy** [comment]
We'll aim to add Jason's fix above to the next release of NCCL

**Comment by hanzhi713** [comment]
Closing the issue since it's fixed in 2.18.3.
